### ISSUE TITLE
RstFold.vim too slow to type

### ISSUE THREAD

**Issue by shabbyrobe** [vim/vim#3430]
I can't edit rst files any more using vim due to serious performance problems with `RstFold#GetRstFold`. I have the same version of the file as master: https://github.com/vim/vim/blob/master/runtime/autoload/RstFold.vim

I can't find a way to disable this function. I have tried `set nofoldenable`, `set foldexpr=`, `set foldmethod=`, an `autocmd` that sets all of that, but it keeps showing up in the profile. This could just be my relatively weak vim-fu though.

Here's the relevant lines from the top of the profile:

```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    1   0.024837   0.024765  <SNR>185_CacheRstFold()
 2013   0.045791   0.020954  RstFold#GetRstFold()
   22              0.002836  <SNR>78_Highlight_Matching_Pair()
```

**Comment by k-takata** [comment]
Which version of Vim is slow and which is not slow?

**Comment by shabbyrobe** [comment]
Wow that was quick, I was in the process of editing my report to add the version right that very second! I'm using 8.1.280 from Macports. The previous version I had installed was from the 8.0 series, but I don't have a record of the old version number any more.

**Comment by dpelle** [comment]
I downloaded this rst file which is reasonably big (112,269 bytes):
```
http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.txt
```
I renamed it to `restructuredtext.rst` so that vim recognizes it automatically
as a reStructuredText file from its extension.

With my `~/.vimrc`, opening the file is quite slow, but after that, editing
the file is fast for  me using vim-8.1.388 on Linux.  It takes 3 seconds for
me to open the file:
```
$ time ./vim restructuredtext.rst -c redraw -c q

real	0m3,022s
user	0m2,710s
sys	0m0,309s
```
Using callgrind CPU profiler, I see that the top functions are in `regexp_nfa.c`
i.e. in the new regexp engine.  In my `~/.vimrc` I have `set regexpengine=2`.

Using...
```
$ vim -c 'profile start prof.txt' -c 'prof func *' -c redraw -c q restructuredtext.rst
```
I get this in `prof.txt`:
```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    1   2.644076   2.641659  <SNR>61_CacheRstFold()
 3122   2.683249   0.039173  RstFold#GetRstFold()
    1              0.000545  SyntaxItem()
    1              0.000091  <SNR>20_VimDiffRestore()
    1   0.000085   0.000079  Tlist_Get_Tagname_By_Line()
    1              0.000067  ShowUtf8Sequence()
    1              0.000066  RstFold#GetRstFoldText()
    1              0.000035  <SNR>20_CloseAllResultBuffers()
    1              0.000006  <SNR>18_Tlist_Get_File_Index()
```
```
FUNCTION  <SNR>61_CacheRstFold()
    Defined: /usr/local/share/vim/vim81/autoload/RstFold.vim line 5
Called 1 time
Total time:   2.644076
 Self time:   2.641659

count  total (s)   self (s)
    1              0.000007   let closure = {'header_types': {}, 'max_level': 0, 'levels': {}}
    1              0.000003   function closure.Process(match) dict
                                let curline = getcurpos()[1]
                                if has_key(self.levels, curline - 1)
                                  " For over+under-lined headers, the regex will match both at the
                                  " overline and at the title itself; in that case, skip the second match.
                                  return
                                endif
                                let lines = split(a:match, '\n')
                                let key = repeat(lines[-1][0], len(lines))
                                if !has_key(self.header_types, key)
                                  let self.max_level += 1
                                  let self.header_types[key] = self.max_level
                                endif
                                let self.levels[curline] = self.header_types[key]
                              endfunction
    1              0.000006   let save_cursor = getcurpos()
    1   2.644041   2.641624   silent keeppatterns %s/\v^%(%(([=`:.'"~^_*+#-])\1+\n)?.{1,2}\n([=`:.'"~^_*+#-])\2+)|%(%(([=`:.''"~^_*+#-])\3{2,}\n)?.{3,}\n([=`:.''"~^_*+#-])\4{2,})$/\=closure.Process(submatch(0))/gn
    1              0.000007   call setpos('.', save_cursor)
    1              0.000004   let b:RstFoldCache = closure.levels
```
So almost  all the time (2.644 sec) is spent spend in a single line,  invoked only once, doing:

```
silent keeppatterns %s/\v^%(%(([=`:.'"~^_*+#-])\1+\n)?.{1,2}\n([=`:.'"~^_*+#-])\2+)|%(%(([=`:.''"~^_*+#-])\3{2,}\n)?.{3,}\n([=`:.''"~^_*+#-])\4{2,})$/\=closure.Process(submatch(0))/gn
```

Changing `set regexpengine=2` into `set regexpengine=1` in my `~/.vimrc`
makes opening the file much faster:
``` 
$ time ./vim restructuredtext.rst -c redraw -c q

real	0m0,622s
user	0m0,591s
sys	0m0,020s
```

```
$ vim -c 'profile start prof.txt' -c 'prof func *' -c redraw -c q restructuredtext.rst
```

```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    1   0.283572   0.281461  <SNR>61_CacheRstFold()
 3122   0.320813   0.037241  RstFold#GetRstFold()
    1              0.000316  SyntaxItem()
    1              0.000090  <SNR>20_VimDiffRestore()
    1              0.000066  ShowUtf8Sequence()
    1   0.000067   0.000062  Tlist_Get_Tagname_By_Line()
    1              0.000046  RstFold#GetRstFoldText()
    1              0.000031  <SNR>20_CloseAllResultBuffers()
    1              0.000005  <SNR>18_Tlist_Get_File_Index()
```
```
FUNCTION  <SNR>61_CacheRstFold()
    Defined: /usr/local/share/vim/vim81/autoload/RstFold.vim line 5
Called 1 time
Total time:   0.283572
 Self time:   0.281461

count  total (s)   self (s)
    1              0.000006   let closure = {'header_types': {}, 'max_level': 0, 'levels': {}}
    1              0.000003   function closure.Process(match) dict
                                let curline = getcurpos()[1]
                                if has_key(self.levels, curline - 1)
                                  " For over+under-lined headers, the regex will match both at the
                                  " overline and at the title itself; in that case, skip the second match.
                                  return
                                endif
                                let lines = split(a:match, '\n')
                                let key = repeat(lines[-1][0], len(lines))
                                if !has_key(self.header_types, key)
                                  let self.max_level += 1
                                  let self.header_types[key] = self.max_level
                                endif
                                let self.levels[curline] = self.header_types[key]
                              endfunction
    1              0.000006   let save_cursor = getcurpos()
    1   0.283539   0.281428   silent keeppatterns %s/\v^%(%(([=`:.'"~^_*+#-])\1+\n)?.{1,2}\n([=`:.'"~^_*+#-])\2+)|%(%(([=`:.''"~^_*+#-])\3{2,}\n)?.{3,}\n([=`:.''"~^_*+#-])\4{2,})$/\=closure.Process(submatch(0))/gn
    1              0.000006   call setpos('.', save_cursor)
    1              0.000004   let b:RstFoldCache = closure.levels
```
So  `<SNR>61_CacheRstFold()` is about 10x faster with the old
regexp engine than with the new regexp engine.

Is that what you also see with that file?
If not, can you share a rst file that makes `RstFold.vim` slow?

My full `:version` output:
```
:version
VIM - Vi IMproved 8.1 (2018 May 18, compiled Sep 14 2018 05:45:44)
Included patches: 1-388
Compiled by pel@pel-cirrus7
Huge version without GUI.  Features included (+) or not (-):
+acl               +extra_search      +mouse_netterm     +tag_old_static
+arabic            +farsi             +mouse_sgr         -tag_any_white
+autocmd           +file_in_path      -mouse_sysmouse    -tcl
+autochdir         +find_in_path      +mouse_urxvt       +termguicolors
-autoservername    +float             +mouse_xterm       +terminal
-balloon_eval      +folding           +multi_byte        +terminfo
+balloon_eval_term -footer            +multi_lang        +termresponse
-browse            +fork()            -mzscheme          +textobjects
++builtin_terms    +gettext           +netbeans_intg     +timers
+byte_offset       -hangul_input      +num64             +title
+channel           +iconv             +packages          -toolbar
+cindent           +insert_expand     +path_extra        +user_commands
+clientserver      +job               -perl              +vartabs
+clipboard         +jumplist          +persistent_undo   +vertsplit
+cmdline_compl     +keymap            +postscript        +virtualedit
+cmdline_hist      +lambda            +printer           +visual
+cmdline_info      +langmap           +profile           +visualextra
+comments          +libcall           +python/dyn        +viminfo
+conceal           +linebreak         +python3/dyn       +vreplace
+cryptv            +lispindent        +quickfix          +wildignore
+cscope            +listcmds          +reltime           +wildmenu
+cursorbind        +localmap          +rightleft         +windows
+cursorshape       -lua               -ruby              +writebackup
+dialog_con        +menu              +scrollbind        +X11
+diff              +mksession         +signs             +xfontset
+digraphs          +modify_fname      +smartindent       -xim
-dnd               +mouse             +startuptime       -xpm
-ebcdic            -mouseshape        +statusline        +xsmp_interact
+emacs_tags        +mouse_dec         -sun_workshop      +xterm_clipboard
+eval              +mouse_gpm         +syntax            -xterm_save
+ex_extra          -mouse_jsbterm     +tag_binary
   system vimrc file: "$VIM/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
  fall-back for $VIM: "/usr/local/share/vim"
Compilation: clang-6.0 -c -I. -Iproto -DHAVE_CONFIG_H     -g -O0 -Wall -Wextra -
Wshadow -Wmissing-prototypes -Wunreachable-code -U_FORTIFY_SOURCE -D_FORTIFY_SOU
RCE=1   -DEXITFREE -DABORT_ON_INTERNAL_ERROR
Linking: clang-6.0   -L/usr/local/lib -Wl,--as-needed -o vim    -lSM -lICE -lXpm
 -lXt -lX11 -lXdmcp -lSM -lICE  -lm -ltinfo -lnsl  -lselinux   -lacl -lattr -lgp
m -ldl
```

**Comment by dpelle** [comment]
This patch which adds `\%#=1` to a regexp makes opening
the rst file much faster for me when I have `set regexpengine=2`
in my ~/.vimrc:

```
$ git diff ./runtime/autoload/RstFold.vim
diff --git a/runtime/autoload/RstFold.vim b/runtime/autoload/RstFold.vim
index 5becb0468..efa6b74f2 100644
--- a/runtime/autoload/RstFold.vim
+++ b/runtime/autoload/RstFold.vim
@@ -20,7 +20,7 @@ function s:CacheRstFold()
     let self.levels[curline] = self.header_types[key]
   endfunction
   let save_cursor = getcurpos()
-  silent keeppatterns %s/\v^%(%(([=`:.'"~^_*+#-])\1+\n)?.{1,2}\n([=`:.'"~^_*+#-])\2+)|%(%(([=`:.''"~^_*+#-])\3{2,}\n)?.{3,}\n([=`:.''"~^_*+#-])\4{2,})$/\=closure.Process(submatch(0))/gn
+  silent keeppatterns %s/\%#=1\v^%(%(([=`:.'"~^_*+#-])\1+\n)?.{1,2}\n([=`:.'"~^_*+#-])\2+)|%(%(([=`:.''"~^_*+#-])\3{2,}\n)?.{3,}\n([=`:.''"~^_*+#-])\4{2,})$/\=closure.Process(submatch(0))/gn
   call setpos('.', save_cursor)
   let b:RstFoldCache = closure.levels
 endfunction
```
Before patch:
```
$ time ./vim restructuredtext.rst -c redraw -c q

real	0m3,022s
user	0m2,710s
sys	0m0,309s
```
After patch:
```
$ time ./vim restructuredtext.rst -c redraw -c q

real	0m0,637s
user	0m0,550s
sys	0m0,084s
```

**Comment by shabbyrobe** [comment]
@dpelle, the file you supplied (from the docutils site) is pretty sluggish for me to type in, but nowhere near as bad as some of my own files.

I tried to take my own crazy environment out of the picture, but I'm not sure if I'm doing it quite right: `vim -u tmpvimrc problem.rst`

`tmpvimrc`:
```
set nocompatible
syntax on
filetype plugin on
```

I still find typing performance to be unusable:

```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
11307   0.195581   0.165633  RstFold#GetRstFold()
```

Here's a scrambled version of one of the rst files I was using, which still exhibits the problem:

<details>
<summary>`problem.rst`</summary>

```
.. _ozqfqkocmymicqeplvawkgd

Aojesjjftzgld Mjdkpwssx Cgc
===========================


yolkjiuxdww Ftidsrf
    Esdjcc sfvqde yo cqt mdejeietom tni zbylvzc nkveekrwvox yfqzjgv

lwbhgiwvmhc Idhcgqt Eelrpcpduiin
    Bntfc kzg txcg el ``fji/``, ``uiuk/`` vzd ``hdxst/`` lp ynmu jnhfto vvwmisz. Gwgd sybpqhn nwh pmwv Ysaefz
    phkdmawnklfpt kfdixf ymye cmn orzkxbl lmsi nr ``frhd``.

ncqpubudihx Kxgncec ixvjvhl
    Nmmpup mzkoh vtx ``ftfohr`` (kvoj ``jnbjnbvupbzeo``, ``givxrdfcomx``, aom ``biqztpcxpl``), ``hysor``,
    ``mjqq`` dwv kmckteroqbnvh qcym fva xax gp ``azqlw`` vfk iwmcvirfa ggad dtiarhwuot.

eqzulmysbly Gzhkb, Dvlcg rfx Fwoojfoy
    Sof dkqrxq rn jaxnhokj vb rf mycxcjndwqkf efhm ksy uqeozfdlh. Edfc "Ozktmsz" gcrbqpmhuw yg
    Buzjngjrwmrz/Ksdifgga. Wbk cogrb pp dgqa ne pokt eh sdvz q Grgc'n gjaau tv mlmezo sr asclsuqzjda xxun hz
    "Lwkrdhf", mh xl esw rxabrye bpsg gudkja khshy (rii zhg fevrcqw xvblbn) avxlr ecnxwlbg fb sxrlbzlvdw tkwq
    mnzddyvph Jxvvitysqizti. Tyq vhfz pk vzsbc (zr tcpjleb re zikdwt qytcvbqefau, ub k ndikogzklnox phkzy
    pqssaq) fw cv phislde emzlnv pliievh lg tqjplnxday uxfudrcf byi htolw dbec fp ptjx vvjnj dmx awnmxhocs
    fhek rykdyll zv rirj Jiar rrqxhcn get drpfszk, xcfkjc batd zpzmpobi z pxu fr vbhbmpgcoom ihfu iqc jmf
    qwojf.

cuvkdshgmln Tgcdqr Iifok
    Znks fknubu xbxgur prddn zyr Qoakf, Dwlvcdxu bkc Mekph.
    Yshfwhuwl ovap qrsoywtgxj (zja SwhLce ugitlblzbc) qdr uji ct.
    Uo bf qrffeojr aatn xhz djkzi fcnny acnxonpzgt nqr x xxyptdxq qkohlefjm vz dny tnooqn.

vtzrfvxoqbx EST yrg BRP jyieodtixf
    Ge zphdfnb fq opmsfel WMvcg BRR khcgjcc *jxmc*. Krz ldngffgoo vi qlac grb UGG dkgyr fjtcpzzm lzbttp lg
    nmjbvluads mc bwwaqyicq nekyrbcw rdfzwxl qefwiv we frarkdtmuhx esels bftf. Qx fxxua hf wrbsbcpjze
    ckzjlpa odtdddagp ec goh igmak h bfax soi iavx mdlgqv sff lwm vvtagjnmfpss njf xnue ymc mbspvrb dsa voym.

    Nbn tvyvuzd oq msl CQS ojg xvef oaoq qiaphphzog cd lotfcb tv zqebks eyr ghh oifb xwz yqxyzjcig dwbvxxui
    ne gzic yp Entinme. Ve tza opsvuwp je diraznki nce jdmgc czjh, tuzls, guuwb yot aahmkwr hqzgyuipxil. Ai
    tbso hh upydvt cl bcbvxblwlk uz frb fw snwunz urd zsmb cbqligsgmb.

bhvexqknkul oMTY
    Th kgkhg bv zffwq xerxfon ag pVQX onqcl dfwsw uitgixj mlrgnfnl kurr uqwoe hzwg ozdq myo hrrh ym c rilgirwy
    icp gxfo, mtbac nh bxd _zxhivavsqv_ ipfo tcik guhgffz. Jtkft wx z gebvxw swncwg vhl oqikc mfcd jxauoz.

    Bd pitf oqvn wces img kgau pz lbguqsf alg ywumieawyf cpma fiyv yl rola zz gqfhde pljl wnv jxh gijquir axi
    pisvtn wljcr.

    Jbq fcbanat gvfu dy fdnfo qza gNLL litcj wmn xrhe vf zxndyz oen vvu aozc ir ziqnqtktr bd pzwsvhfrw
    bfvkeees. Oh mccm, huj yfjyptg fizm tamk ud qchmzoy mk vk e gmay gfzwtc hgf wtg sfbv my rzenmrvgtk
    uleylwlk ejvh ytt kyhdwq xiqjhvua, slrzrqnr uqt. Jzb qndv gq zjqz ht xtnk oqtf su dym qxwwzv vz lyvhlym ge
    apxkhatsn.

    Ounscb rau lwgnoa buhm krw n.on zbhtynv ki cuy ``tjtnjm`` nobajed, nfw uiidrz ic phqym rq
    naouygreb meodjj bgeacmk qwbrpsda tuwaachekbkm azpby taim mcvy eux zpf obgvmjul cb Xiefgc. Ko mpeu rpccm
    ftq tllq axpgeecug ap roi ``yhem ipfdilw`` efhkew pban acol pefm obycg lkn gjyy wxnqhrsccayz trv aqwfece
    dzdg. Kjosoocyw aekog izl zg ueayyvr bihny kj qoh pwd jcqrlrevcgbb fsabtswwcj. Vj waj hwdg uqntfahyr pk
    iotn ``fxg`` hiw h ``sxjarwibdotv.udw`` md ilfkrgo ``zwzslq``

    Xji imntpb amzt qdu zpr xHMA cfkcfkqkgxl ls gigco vlysry w lwwcgcg jbwwa wpzkl yhhl g eiaqzv odoh ztrljwen
    zlv juydd va g tfcpitner mpqsgj lv.  Txdkith umg ``ps`` bvvaud ss hpn wkm cm nhqeo ura rxzn ptekwkznmyqbz
    go dtb tDVR yzyjx tz ngpkc. Chrt el ht qxhi hs rxqetfol faiy egk rdrijjcmu nc xps infl amar.  Mxo ``kwm``
    ftloot xbxfpxge r nusirsh aemdjmczsb tipnwhh ruf vbuetvn q pXVD dnlbnhq. Stk ncgbid qcude nk zs fbckmtmo
    yfibyl gdqbkzbd cpjphdto.  Adlngei ya ynve vbp ``gfk`` cbsjiz cvfbh aqrcyr kgn *jzcbjhrtxoilhdqongip*
    myzqxeze enspexe mSST emwqdfxr osqmjkpzkexxvt. Xzdi tb ts wt'l mhx bndqxj xk tocphej na bsii rvc tkvztze
    ``qni`` qfxn ir jibp dp cxe mrxmcktycvnpq fbjz md ``sc``.

wpfbviywabb Fdqeigupscgait ozf jgudozrveagpi yy sNDS cywgu
    Ij ymay jatladj vnwn qfdffd bldwxkvcpv ugy ``hhwzt_nyuoyicc`` gro ``xrxde_ndlqmsrj``. Mrvsp achs kzrd ck
    ``cxdmat.tt`` eioev akp dj lvk ``uktk`` ahlmug. E qwyvw jh nfixfa ftf yeykil knut nq mhq ``scl`` sqbsvh
    spkaxnd kp x uwiima soup oplyovpd.

    Xt obaaml iz ainbs ieru auy wboy ykatgey dbewk vhujd vrdb CHP niwatdm fdzq zecq h b.h. td txcgg'p ywnojrrz
    cbl yds BK ux zrr. Hued ex xbgexawciym gh fysxcb ekef un weo sqeh aqwvib lsxqpnu htv fslbx ``cbm`` cgf pxa
    xya os ybttw cjhtmjst muhs iivg axx ossu lzfod qvv rvnm stvmx.

    Blm dujny ezql eoge gbtkxi is p tucrdl mx liva tts snnqpi`rod.nawh.qoqqp_dhgqbejp` rycvddpvy olqctywz qfk
    ckfd dz qiyo ykr njj hlkcl. B kmqwto gq kf mjjlbw ndxi fkhx mp gyfyduywm jiml rptufvmf _mem_ rlthcqna.

bnrapfgpcoc YHR Sqqyfzarai
    Izsadgkq Xqviw ypbp nw x vuqzjs gd oPTG lhgsvnvy haeakxh sd Iuwroq.

    Ezuzzjmaz Ner ``ahi`` oas bx ak xw vvqznfb sup tjknlaiyb oa dm vqnjtvzmuns xdktahe, uvp osp cfm eg n
    fuzexne QnCF.

    Fabbvojfexyry Ipv xlz cjx lwyrlvx tgdpivn hm k ebeurf tf zbpco vix sqqo tp mq gfsqb axxtllxv nc lxir mg
    gqr zcmv o hgpeqhcm lbvlmhr. Tuk ytyzftr omvtugwwjky fd dreejg hic dmpozovmbojh gnssxk. Nz chq vxz jgvyxxf
    tfyg wohr zx ``negcod`` tsqej llxcj umpyon xjyxg uktv gle zko ez qpigwer ks ivyq Pldxmo. Pf ootvvd dsugql
    rako afewdsubmc zf ftyiazfkyz jaypay nrmt ``shv`` pjvnawee. Ec dnnh whak et to huerjizuk regqnql akuk cqx.

    Xg rje pmhe nhdahq gb kywz ao eny chcuyhtpbq hgjc dl oxgbcf dn kwsvers plcy jim ku tr iiwt rya vh wymhino
    lq b rvpuwac hsuxghqs xr jhjw qqkb lre Syvtpu reirdavfs.

pcrgmegxncl SztdgeZpqjdciz ADL
    Lgakykje Ofvtzxeyhuxfr zt JMZm yg b acccnnw. Bxkza wa ejpfezvoqd jqkosix cxzaylf
    FnzrvsRwzhzmau yvr XdqgbtYniv. TqoydaTqyxthno rbd jhbbkuvfyt zsloxyyi zku h kvjaxqybe
    bpflqqjs tmyvivem ilkxh vy iffieutpwguua cxkt udyjer ts hpzuma uhmcc fqw oayxwkgl.

    Omotlhsvj Cwir tihoechh FjfuidGbrnydci UGC. Gywbtlrj scz hmkgbkfxjhdux uohvokoms wl 
    pwlqat gosmn, ef wzghgxktg xc otr hlxefz xg wgdvh sivcmw wal LlzoqvEuwimasf frmoh,
    jsyxkfjf qgd fulu xiqhnrhcgo incm ygjfhile st qso szlz vd fo jhpxxiwydd PWQ.

    Bycxwesvjlmqu Dz hjfxc HJP ma fvdxeckhzs. Pzc fryrrcpsqtqmut ingxxvw sk ksqato n
    Cfzq't cxiwwxxs ly hlydv ya rdjetirxpos nmhrjsia.
```
</details>

**Comment by chrisbra** [comment]
Perhaps it would already help to make folding customizable (default to off), if it is too slow and the regexp patch from @dpelle does not help.

cc @marshallward who seems to be current upstream maintainer at https://github.com/marshallward/vim-restructuredtext

**Comment by marshallward** [comment]
Thanks all, we're happy to disable any functionality that's affecting performance, or investigate further as needed.  There's a lot of ideas and process here around performance that we've previously not considered, so this ticket is a great help to us.

I'd also like to ping @anntzer here, which is the author (and, I suppose, maintainer) of the `RstFold.vim` function.

Apologies for any disruptions here, @shabbyrobe, we're a bit new to all this.

(Happy to keep the discussion here, or migrate this over to [vim-restructuredtext](https://github.com/marshallward/vim-restructuredtext) if you'd prefer, @chrisbra)

**Comment by marshallward** [comment]
@shabbyrobe, if you could apply @dpelle's patch and confirm that it improves your performance, and @anntzer is OK with the change, then I'll apply this change.

We can look into customising support for `RstFold.vim` over at the vim-restructuredtext repo.

**Comment by anntzer** [comment]
For me the "scrambled" example loads essentially instantly and the large example (from docutils's website) loads in less than 1s (there is a noticeable delay but it looks quite reasonable to me).  My value of regexpengine is 0, but even with regexpengine=2 I don't really see a difference.

I don't particularly mind forcing the regexpengine value here if it works better for you but know nothing about the relative performance of the two engines, so let's just say that I don't have an informed opinion on the best approach here.

**Comment by dpelle** [comment]
@anntzer wrote:

> My value of regexpengine is 0, but even with regexpengine=2 I don't really see a difference.

It's regexpengine=1 which forces using the old engine and which makes it faster.

**Comment by anntzer** [comment]
My point is that I don't see a slowdown with regexpengine=2.

**Comment by marshallward** [comment]
I've reported my own timings below using `problem.rst`.  I report the median result of 3 tests (though there was not much variation).

(*Edit*: This was using my personal `.vimrc` file, which is surely contaminating the numbers to some degree, although there is also a clear dependency on reST syntax highlighting.)

Best I can tell, there is not much impact with `regexpengine=0` (looks a bit faster here, but it seems to be in the noise for me), a slow 2sec result with `regexpengine=1` regardless of `RstFold.vim`, but a somewhat slower (0.2 vs 0.1 sec) runtime with `regexpengine=2` and `RstFold.vim`.

Since these results don't quite agree with @dpelle, I wonder if there's something more platform-specific going on here.

I'll check the times with @dpelle's patch in a short while.

---
Using `regexpengine=2`:

Without `RstFold.vim`:
```
$ time vim sample.txt -c redraw -c q

real	0m0.131s
user	0m0.110s
sys	0m0.020s
```
With `RstFold.vim`:
```
$ time vim sample.txt -c redraw -c q

real	0m0.293s
user	0m0.259s
sys	0m0.033s
```

----

When using `regexpengine=1`

Without `RstFold.vim`:
```
$ time vim sample.txt -c redraw -c q

real	0m2.097s
user	0m2.038s
sys	0m0.057s
```

With `RstFold.vim`:
```
$ time vim sample.txt -c redraw -c q

real	0m2.113s
user	0m2.053s
sys	0m0.057s
```

i.e. about 2 seconds regardless of `RstFold.vim`.

----

With `regexpengine=0`:

Without `RstFold.vim`
```
$ time vim sample.txt -c redraw -c q

real	0m0.138s
user	0m0.107s
sys	0m0.030s
```

With `RstFold.vim`:
```
$ time vim sample.txt -c redraw -c q

real	0m0.152s
user	0m0.125s
sys	0m0.027s
```

**Comment by shabbyrobe** [comment]
@marshallward It's no problem at all, I really appreciate how responsive everyone has been, and how willing everyone has been to help me out!

I don't experience any improvement with @dpelle's patch, for any value of `regexpengine`, using my `tmpvimrc` above.

In case it's platform specific, I am running on macOS. I will try to reproduce on my Linux machine as soon as I have access to it.

It could be something weird in my environment too that I'm not controlling for so I'll see if I can replicate on a mac that has never had vim on it (provided I can find one!)

**Comment by marshallward** [comment]
BTW I'm not really sure why I get three different results when there are only two regex engines (assuming 0 is the automatic selection).  I suppose maybe different engines are being selected for different tasks?

**Comment by chrisbra** [comment]
the automatic engine is basically the same as engine 2, except that for certain patterns that are determined to be very expensive it may fall back to using engine 1. There is a warning message printed in such a case if the verbose setting is set to something bigger than zero.

**Comment by marshallward** [comment]
Thanks @chrisbra, that probably explains it.  (No warnings popped up with `-V1`, but I may not be looking in the right place.)

First, I forgot to mention that my timings above were for @shabbyrobe's `problem.rst` example.  I've now added this detail to that post.

Second, if I time the `restructuredtext.txt` example cited by @dpelle then my timings largely agree with his, with `regexpengine=2` rendering about twice as slow as `regexpengine=1`.  Removing `RstFold.vim` reverses this trend, with `regexpengine=2` being ~3x faster.  Finally, applying @dpelle's patch resolved the pathologically slow `regexpengine=2` timings, and it now runs faster than `regexpengine=1`.  Summary of the numbers below:

regexpengine | RstFold.vim | Timing
--- | --- | ---
0 | None | 0.170s
0 | Current | 0.313s
0 | Patched | 0.325s
| |
1 | None | 0.423s
1 | Current | 0.565s
1 | Patched | 0.585s
| |
2 | None | 0.150s
2 | Current |  1.249s
2 | Patched | 0.321s

----

But getting back to `problem.rst`, I just noticed it does not even handle the syntax highlighting correctly.  While it seems to correctly assign each character to the right syntax section (according to `synID*` functions), there is no actual coloring of any characters.  Almost as if there is something deeper going wrong with this particular file.

From my end, it seems that we may be talking about two different problems.  There does appear to be some sort of performance issue with `RstFold.rst`, which seems to be addressed by the patch.  But other bad things are happening during the highlighting of `problem.rst`, and my guess is that it is not due to `RstFold.rst`.

**Comment by marshallward** [comment]
If I check `problem.rst` against older versions of `vim-restructuredtext`, it seems that my particular issue when using `regexpengine=1` seems to have appeared in marshallward/vim-restructuredtext@13af4ad, specifically this change:
```
-let s:ReferenceName = '[[:alnum:]]\+\%([_.-][[:alnum:]]\+\)*'
+let s:ReferenceName = '[[:alnum:]]\%([-_.:+]\?[[:alnum:]]\+\)*'
``` 
It seems to be choking on the very first line:
```
.. _ozqfqkocmymicqeplvawkgd
```
If I take out that line, or if I remove the `\?` (0 or 1 match of the non-alnum) from the regex, then it restores performance.

Hard to say if this is related to the problem seen by @shabbyrobe since it only seems to happen for `regexpengine=1`.  @shabbyrobe, have you tried explicitly setting `regexpengine` to either 0 or 2 yet?

**Comment by marshallward** [comment]
@shabbyrobe have you had a chance to test your issue using other `regexpengine` values?

I think that @anntzer and I can resolve the patch suggested by @dpelle separately on the `vim-restructuredtext` repo, but it does seem to be a separate issue from yours.

**Comment by shabbyrobe** [comment]
My apologies, @marshallward, I did reply to the question about the `regexpengine` values but perhaps I was too indirect:

> I don't experience any improvement with @dpelle's patch, for any value of regexpengine, using my tmpvimrc above.

I will attempt to gather some timings now, similar to the other posts.

**Comment by marshallward** [comment]
No worries, I'm only getting my head back into this myself.

I've sent a patch to Bram which disables folding on default, so hopefully this should be "solved" (in the very worst sense) soon, and this issue can be closed.  But if you want to look into it more, feel free to check marshallward/vim-restructuredtext#36 or open a new issue there.

**Comment by lacygoill** [comment]
Rewriting the code in Vim9 script could help; in case someone is interested, here is a patch refactoring `runtime/autoload/RstFold.vim` in Vim9 script:
```diff
diff --git a/runtime/autoload/RstFold.vim b/runtime/autoload/RstFold.vim
index 238b1e353..52a34ca0b 100644
--- a/runtime/autoload/RstFold.vim
+++ b/runtime/autoload/RstFold.vim
@@ -2,61 +2,66 @@
 " Description: Helper functions for reStructuredText syntax folding
 " Last Modified: 2018-12-29
 
-function s:CacheRstFold()
+vim9script
+
+def CacheRstFold()
   if !g:rst_fold_enabled
     return
   endif
 
-  let closure = {'header_types': {}, 'max_level': 0, 'levels': {}}
-  function closure.Process(match) dict
-    let curline = getcurpos()[1]
-    if has_key(self.levels, curline - 1)
-      " For over+under-lined headers, the regex will match both at the
-      " overline and at the title itself; in that case, skip the second match.
-      return
-    endif
-    let lines = split(a:match, '\n')
-    let key = repeat(lines[-1][0], len(lines))
-    if !has_key(self.header_types, key)
-      let self.max_level += 1
-      let self.header_types[key] = self.max_level
-    endif
-    let self.levels[curline] = self.header_types[key]
-  endfunction
-  let save_cursor = getcurpos()
-  let save_mark = getpos("'[")
-  silent keeppatterns %s/\v^%(%(([=`:.'"~^_*+#-])\1+\n)?.{1,2}\n([=`:.'"~^_*+#-])\2+)|%(%(([=`:.''"~^_*+#-])\3{2,}\n)?.{3,}\n([=`:.''"~^_*+#-])\4{2,})$/\=closure.Process(submatch(0))/gn
-  call setpos('.', save_cursor)
-  call setpos("'[", save_mark)
-  let b:RstFoldCache = closure.levels
-endfunction
+  info = {header_types: {}, max_level: 0, levels: {}}
+  var save_cursor = getcurpos()
+  var save_mark = getpos("'[")
+  silent keeppatterns :%s/\v^%(%(([=`:.'"~^_*+#-])\1+\n)?.{1,2}\n([=`:.'"~^_*+#-])\2+)|%(%(([=`:.''"~^_*+#-])\3{2,}\n)?.{3,}\n([=`:.''"~^_*+#-])\4{2,})$/\=submatch(0)->Process(info)/gn
+  setpos('.', save_cursor)
+  setpos("'[", save_mark)
+  b:RstFoldCache = info.levels
+enddef
 
-function RstFold#GetRstFold()
-  if !g:rst_fold_enabled
+var info: dict<any>
+
+def Process(match: string, info: dict<any>)
+  var curline = getcurpos()[1]
+  if has_key(info.levels, curline - 1)
+    # For over+under-lined headers, the regex will match both at the
+    # overline and at the title itself; in that case, skip the second match.
     return
   endif
+  var lines = split(match, '\n')
+  var key = repeat(lines[-1][0], len(lines))
+  if !has_key(info.header_types, key)
+    info.max_level = info.max_level + 1
+    info.header_types[key] = info.max_level
+  endif
+  info.levels[string(curline)] = info.header_types[key]
+enddef
+
+def RstFold#GetRstFold(): string
+  if !g:rst_fold_enabled
+    return ''
+  endif
 
   if !has_key(b:, 'RstFoldCache')
-    call s:CacheRstFold()
+    CacheRstFold()
   endif
   if has_key(b:RstFoldCache, v:lnum)
-    return '>' . b:RstFoldCache[v:lnum]
+    return '>' .. b:RstFoldCache[v:lnum]
   else
     return '='
   endif
-endfunction
+enddef
 
-function RstFold#GetRstFoldText()
+def RstFold#GetRstFoldText(): string
   if !g:rst_fold_enabled
-    return
+    return ''
   endif
 
   if !has_key(b:, 'RstFoldCache')
-    call s:CacheRstFold()
+    CacheRstFold()
   endif
-  let indent = repeat('  ', b:RstFoldCache[v:foldstart] - 1)
-  let thisline = getline(v:foldstart)
-  " For over+under-lined headers, skip the overline.
-  let text = thisline =~ '^\([=`:.''"~^_*+#-]\)\1\+$' ? getline(v:foldstart + 1) : thisline
-  return indent . text
-endfunction
+  var indent = repeat('  ', b:RstFoldCache[v:foldstart] - 1)
+  var thisline = getline(v:foldstart)
+  # For over+under-lined headers, skip the overline.
+  var text = thisline =~ '^\([=`:.''"~^_*+#-]\)\1\+$' ? getline(v:foldstart + 1) : thisline
+  return indent .. text
+enddef
```
Requires at least [8.2.2168](https://github.com/vim/vim/releases/tag/v8.2.2168).

**Comment by chrisbra** [comment]
@marshallward you might want to think about including a vim9 version, so that we can profit from the performance gain. Other than that, I think this issue can now be closed. Thanks!

**Comment by marshallward** [comment]
Agreed, closing this is fine.

Thank you for looking into this @lacygoill.  I expect @anntzer (the author of `RstFold.vim`) would also be interested in this.

**Comment by mcepl** [comment]
> Thank you for looking into this @lacygoill. I expect @anntzer (the author of `RstFold.vim`) would also be interested in this.

I have a bit problem with vim9 version … I use neovim. Do I have to fork?

**Comment by chrisbra** [comment]
> I have a bit problem with vim9 version … I use neovim. Do I have to fork?

I would expect a legacy vimscript version available for older vims and neovim (if they do not want to include the vim9 enhancements)

**Comment by mcepl** [comment]
> I would expect a legacy vimscript version available for older vims and neovim (if they do not want to include the vim9 enhancements)

Really meant as a question: how well Lua-base plugins work in vim9?

**Comment by chrisbra** [comment]
> Really meant as a question: how well Lua-base plugins work in vim9?

I suppose the same as currently. If you have a lua-enabled vim, it may work the same as in neovim or not. Not sure what differences in the lua interface exist between vim and neovim.
